### ISSUE TITLE
Plane stress/strain and constitutive laws

### ISSUE THREAD

**Issue by Psirus** [nutofem/nuto#13]
Currently, the info whether you want plane stress or plane strain behaviour is in the section object, attached to the element in `mSection`. If a constitutive law needs this information, it gets it through the element pointer, i.e. `rElement->GetSection()->GetType()`. This is to be avoided, because we want the constitutive laws independent from the element. This issue is to discuss how to handle this going forward. Ideas so far:
## Add directly into input list

In the evaluate function of the element, you check whether the element is 2D, and if it is, add this information (as a new constitutitve input type) into the input map, e.g. `constitutiveInput[Constitutive::Input::PLANE_BEHAVIOUR] = std::make_unique<ConstitutivePlaneBehaviour>();`. 
Pro: 
- relatively straightforward 

Con: 
- not really an input (as opposed to, say, strain)
- duplication (behaviour stored at section, and then "copied" into input list)
## As parameter of the constitutive law

Treat it like a parameter, e.g. Young's modulus. Once you create the law, you'll also need to call e.g. `structure.ConstitutiveLawSetParameterBool(law_id, true)`.
Pro:
- no need to manipulate the input list of the law

Con:
- you'd need two constitutive laws for structures where you have both plane stress and plane strain elements, and assign the laws accordingly
- not really a parameter either (conceptually ugly)
## Have EngineeringStress<2>/EngineeringStrain<2> contain this information

Let the engineering strain and stress have this information, so that a) the constitutive law does something different if it is supposed to calculate e.g. a plane stress EngineeringStress, and b) calculations using these objects take advantage of this knowledge, i.e. the results depend on the "type" of stress you're using, and you don't need to explicitly handle the two cases everywhere.
Pro:
- could possibly be quite an elegant, "polymorphic" solution (doesn't necessarily need inheritance/templates)

Con:
- not straightforward to implement (at least for me)
## Something else?

I'm really not sure which way to go. None of the suggestions strike me as "There should be one-- and preferably only one --obvious way to do it", which is one of the ways I usually try when reasoning about design. Where would you categorize this plane stress/strain behaviour: an element information, a section information, an constitutive parameter or input? Capturing what is the "this is where I would expect it to be, if I knew nothing about NuTo" location would be my preference.

**Comment by TTitscher** [comment]
TL;DR: ConstitutiveLaw
I've been thinking about that for a while now, not that easy. I thought: Who cares about the plane stress/strain state, meaning whose behavior is influenced by that. Say, from a design point of view. 
The section does not care, the element neither. At the moment, the only classes that contain something like switch/if (PlaneState) are the constitutive laws. And thats where I would store it. 

If you now want to implement it, you have to add it in every constitutive law that has something to do with stress. And you have to add it in all laws in the same way. And in every class you have to a SetParameter / GetParameter case. This clearly violates design princliples, like [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourselfl). But that might be a topic for another issue. :)

**Comment by joergfunger** [comment]
This is only partially true, since axisymmetric formulations are another
option that influences the element behavior.

Jörg F.

Am 27.10.2016 8:50 PM schrieb "TTitscher" notifications@github.com:

**Comment by TTitscher** [comment]
I think, we should decide whether we want to base current design decisions on (possible) future implementations or not. Or to which extent.
1. In the _current_ version of NuTo, there are no axisymmetric elements. So I think it is OK to put the plane state at the constitutive law. We can try to implement it with the least amount of duplicate code, test the results and so on. 
2. What we can do, is to guess how a axisymmetric element could be implemented. Two problems here:
   1. The developer has to be familiar with the model. In the case of axisymmetric elements, I was not. (But I had a [quick look](http://www.oofem.org/resources/doc/elementlibmanual/html/elementlibmanualsu7.html))
   2. The developer has to rely on the fact that NuTo will not change a lot until the actual implementation. 

So if I wanted to implement these functionality right now, I would probably derive a new AxisymmetricElement from ContinuumElement<3> and override the N and B matrices to correctly calculate the four stress components [s_x, s_y, s_theta, s_xy]. This would not depend on the plane state. 

@joergfunger, I think you had a different implementation in mind. Maybe by adding "axisymmetric" to the plane state, so that the plane state contains (plane stress, plane strain, axisymmetry). And then change the ContinuumElement accordingly.

And maybe, no one ever actually needs an axisymmetic element... 

I would like to stick to the first principle: Work/rely on the code that is actually there. How do you think about that?

(And I am sorry, that this drifts away from pure discussion about the plane state. But it still meets the "design" label...)

**Comment by Psirus** [comment]
Solved in PDE.
